### ISSUE TITLE
Inference of TypeVar does not work with singledispatch

### ISSUE THREAD

**Issue by GuillaumeDesforges** [microsoft/pyright#3471]
**Describe the bug**  
When using singledispatch with a TypeVar, a type issue is reported even though the type hints are okay.

**To Reproduce**  
Write the code below, run pyright.

**Expected behavior**  
There should be no type issue.

**Screenshots or Code**

```python
from functools import singledispatch
from typing import Any, Callable, Generator, TypeVar

import numpy as np

T = TypeVar("T")


def some_func(
    x,
    f: Callable[[Any], T] = np.array,
) -> Generator[T, None, None]:
    yield f(x)


# typechecks
x: list[np.ndarray] = list(some_func(0))


@singledispatch
def some_generic_func(
    x,
    f: Callable[[Any], T] = np.array,
) -> Generator[T, None, None]:
    raise NotImplementedError


@some_generic_func.register(np.ndarray)
def _(
    x: np.ndarray, f: Callable[[np.ndarray], np.ndarray] = lambda x: x
) -> Generator[np.ndarray, None, None]:
    yield f(x)


# does not typecheck
x: list[np.ndarray] = list(some_generic_func(np.array(0)))
```

reports

```
Expression of type "list[T@some_generic_func]" cannot be assigned to declared type "list[ndarray[Unknown, Unknown]]"
  TypeVar "_T@list" is invariant
    Type "T@some_generic_func" cannot be assigned to type "ndarray[Unknown, Unknown]"
```

**VS Code extension or command-line**  
Pylance v2022.5.1

**Comment by erictraut** [comment]
TypeVars are "solved" when calling a function. In the first part of your sample, you are calling `some_func`, and the type variable `T@some_func` is solved as `np.array`.

In the second part of your example, the decorator `singledispatch` is called with an argument of `some_generic_func`, and the return type of this call is `_SingleDispatchCallable[Generator[T@some_generic_func, None, None]]`. The problem is that there is no way to "solve" the TypeVar `T@some_generic_func` at this point because it has been divorced from its original function.

So I think pyright is doing the right thing here. I don't consider this a type checker bug.

One could make the argument that an unsolved — and unsolvable — TypeVar should take on the type of `Unknown`. That would at least silence the error, but it's probably not what you intended.

Determining when a TypeVar is no longer solvable is difficult for a type checker to determine given the complex (and not completely specified) scoping rules associated with TypeVars, but I'll give some more thought to whether pyright could make that determination in this case.

**Comment by GuillaumeDesforges** [comment]
Thanks for your explanation.

Would it make more sense to update the type stub?

https://github.com/python/typeshed/blob/dbccdcb50e8a2d436b871719c141ac59514aed1a/stubs/singledispatch/singledispatch.pyi#L16

Would you see a way to change it so that the typevar could solve?

**Comment by erictraut** [comment]
I don't see an obvious way to modify the type stub. 

Here's a simplified example that exhibits the problem:
```python
from functools import singledispatch
from typing import Any, Callable, TypeVar

T = TypeVar("T")

@singledispatch
def func(x: Any, f: Callable[[Any], T]) -> T: ...

x: int = func(0)
```

**Comment by erictraut** [comment]
Looking at this in more detail, I think pyright is doing the right thing here.

Mypy has extensive special-case logic to handle `functools.singledispatch`, but without such special-case logic, it produces the same errors as pyright in this case.

For now, I'm going to close this issue because I don't consider it a bug.

We would consider adding special-case logic for `functools.singledispatch` in pyright if there is sufficient demand from pyright users. If this is something you would like to see, please upvote this issue (i.e. add a thumbs-up reaction).

**Comment by GuillaumeDesforges** [comment]
Thanks for investigating this. It makes sense that this becomes a "feature request" :ok\_hand: 

Doesn't closing the issue make it harder to discover? Do you mind if I open a new issue describing a proper feature request?

**Comment by erictraut** [comment]
You can open another feature request if you think it will help with discoverability, but I will likewise close it. I don't keep feature requests open if there is no plan to implement the feature. We can re-open it if there is sufficient interest at some point in the future.

**Comment by GuillaumeDesforges** [comment]
How many likes would "sufficient interest" be for you? :slightly\_smiling\_face:

**Comment by robertmaxton42** [comment]
> You can open another feature request if you think it will help with discoverability, but I will likewise close it. I don't keep feature requests open if there is no plan to implement the feature. We can re-open it if there is sufficient interest at some point in the future.

This seems both self-perpetuating and easy to misunderstand.

**Comment by kevinleahy-switchdin** [comment]
fyi if anyone comes across this, feature request for implementing this is here: https://github.com/microsoft/pyright/issues/4472

Plz upvote if it's relevant to you.
